### PR TITLE
snapshot: upgrade lock usage

### PR DESCRIPTION
**Pull request by mask-pp**

### CODE REVIEW

File: core/state/snapshot/difflayer.go [hunk 1]
```diff
@@ -525,19 +525,20 @@ func (dl *diffLayer) StorageList(accountHash common.Hash) ([]common.Hash, bool)
 		dl.lock.RUnlock()
 		return list, destructed // the cached list can't be nil
 	}
+	storageMap := dl.storageData[accountHash]
 	dl.lock.RUnlock()
 
-	// No old sorted account list exists, generate a new one
```
Comment (rjl493456442) [inline core/state/snapshot/difflayer.go:530]: I don't believe we will achieve any performance improvement, and this change will make the code harder to understand. Please revert it.

File: core/state/snapshot/snapshot.go [hunk 2]
```diff
@@ -818,17 +818,21 @@ func (t *Tree) Verify(root common.Hash) error {
 // The lock of snapTree is assumed to be held already.
 func (t *Tree) disklayer() *diskLayer {
 	var snap snapshot
+	t.lock.RLock()
```
Comment (rjl493456442) [inline core/state/snapshot/snapshot.go:821]: The description said the snap tree lock is assumed to be held here. Please revert this change.

However, i do find in one place we forget to add the lock, please fix it there.

```go
// Release releases resources
func (t *Tree) Release() {
	t.lock.RLock()
	defer t.lock.RUnlock()

	if dl := t.disklayer(); dl != nil {
		dl.Release()
	}
}
```

File: core/state/snapshot/snapshot.go [hunk 3]
```diff
@@ -818,17 +818,21 @@ func (t *Tree) Verify(root common.Hash) error {
...
+	t.lock.RUnlock()
 	if snap == nil {
 		return nil
 	}
 	switch layer := snap.(type) {
 	case *diskLayer:
 		return layer
 	case *diffLayer:
+		layer.lock.RLock()
+		defer layer.lock.RUnlock()
```
Comment (rjl493456442) [inline core/state/snapshot/snapshot.go:835]: good catch.

File: core/state/snapshot/difflayer.go (hunk 1 above)
Comment (mask-pp) [reply core/state/snapshot/difflayer.go]: Thanks for the explanation, and reverted. :)

File: core/state/snapshot/snapshot.go (hunk 2 above)
Comment (mask-pp) [reply core/state/snapshot/snapshot.go]: reverted!

**Review by MariusVanDerWijden — APPROVED** [review]
LGTM,

I created the following test to show this issue:
```go
// Tests that concurrent accesses to the disklayer do not race with each other
func TestDiskLayerRace(t *testing.T) {
	// setAccount is a helper to construct a random account entry and assign it to
	// an account slot in a snapshot
	setAccount := func(accKey string) map[common.Hash][]byte {
		return map[common.Hash][]byte{
			common.HexToHash(accKey): randomAccount(),
		}
	}
	// Create an empty base layer and a snapshot tree out of it
	base := &diskLayer{
		diskdb: rawdb.NewMemoryDatabase(),
		root:   common.HexToHash("0x01"),
		cache:  fastcache.New(1024 * 500),
	}
	snaps := &Tree{
		layers: map[common.Hash]snapshot{
			base.root: base,
		},
	}
	// 4 layers in total, 3 diff layers and 1 disk layers
	snaps.Update(common.HexToHash("0xa1"), common.HexToHash("0x01"), nil, setAccount("0xa1"), nil)
	snaps.Update(common.HexToHash("0xa2"), common.HexToHash("0xa1"), nil, setAccount("0xa2"), nil)
	snaps.Update(common.HexToHash("0xa3"), common.HexToHash("0xa2"), nil, setAccount("0xa3"), nil)
	// Write the disklayer in a different thread
	go snaps.Cap(common.HexToHash("0xa3"), 2)
	// Release the tree
	snaps.Release()
}
```

Which fails as expected on master when running `go test --race` 

```
==================
WARNING: DATA RACE
Write at 0x00c000018090 by goroutine 2258:
  runtime.mapassign()
      /home/matematik/sdk/go1.22.1/src/runtime/map.go:579 +0x0
  github.com/ethereum/go-ethereum/core/state/snapshot.(*Tree).cap()
      /home/matematik/go/src/github.com/ethereum/go-ethereum/core/state/snapshot/snapshot.go:496 +0x25c
  github.com/ethereum/go-ethereum/core/state/snapshot.(*Tree).Cap()
      /home/matematik/go/src/github.com/ethereum/go-ethereum/core/state/snapshot/snapshot.go:419 +0x1ed
  github.com/ethereum/go-ethereum/core/state/snapshot.TestDiskLayerRace.gowrap1()
      /home/matematik/go/src/github.com/ethereum/go-ethereum/core/state/snapshot/snapshot_test.go:518 +0x67

Previous read at 0x00c000018090 by goroutine 2257:
  runtime.mapiterinit()
      /home/matematik/sdk/go1.22.1/src/runtime/map.go:816 +0x0
  github.com/ethereum/go-ethereum/core/state/snapshot.(*Tree).disklayer()
      /home/matematik/go/src/github.com/ethereum/go-ethereum/core/state/snapshot/snapshot.go:821 +0x84
  github.com/ethereum/go-ethereum/core/state/snapshot.(*Tree).Release()
      /home/matematik/go/src/github.com/ethereum/go-ethereum/core/state/snapshot/snapshot.go:669 +0x26
  github.com/ethereum/go-ethereum/core/state/snapshot.TestDiskLayerRace()
      /home/matematik/go/src/github.com/ethereum/go-ethereum/core/state/snapshot/snapshot_test.go:519 +0x817
  testing.tRunner()
      /home/matematik/sdk/go1.22.1/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      /home/matematik/sdk/go1.22.1/src/testing/testing.go:1742 +0x44

Goroutine 2258 (running) created at:
  github.com/ethereum/go-ethereum/core/state/snapshot.TestDiskLayerRace()
      /home/matematik/go/src/github.com/ethereum/go-ethereum/core/state/snapshot/snapshot_test.go:518 +0x80a
  testing.tRunner()
      /home/matematik/sdk/go1.22.1/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      /home/matematik/sdk/go1.22.1/src/testing/testing.go:1742 +0x44

Goroutine 2257 (running) created at:
  testing.(*T).Run()
      /home/matematik/sdk/go1.22.1/src/testing/testing.go:1742 +0x825
  testing.runTests.func1()
      /home/matematik/sdk/go1.22.1/src/testing/testing.go:2161 +0x85
  testing.tRunner()
      /home/matematik/sdk/go1.22.1/src/testing/testing.go:1689 +0x21e
  testing.runTests()
      /home/matematik/sdk/go1.22.1/src/testing/testing.go:2159 +0x8be
  testing.(*M).Run()
      /home/matematik/sdk/go1.22.1/src/testing/testing.go:2027 +0xf17
  main.main()
      _testmain.go:147 +0x2bd
==================
--- FAIL: TestDiskLayerRace (0.00s)
    testing.go:1398: race detected during execution of test
FAIL
exit status 1
FAIL    github.com/ethereum/go-ethereum/core/state/snapshot     8.698s
```
